File: remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        curr = head

        for i in range(n):
            curr = curr.next

        if curr == None:
            head = head.next
            return head

        prev = head

        while curr.next != None:
            curr = curr.next
            prev = prev.next

        prev.next = prev.next.next

        return head

if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)

    s = Solution()
    head = s.removeNthFromEnd(head, 2)

    curr = head

    while curr != None:
        print(curr.val)
        curr = curr.next

File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 */
function ListNode(val) {
  this.val = val;
  this.next = null;
}
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
  let fast = head;
  let slow = head;

  for (let i = 0; i < n; i++) {
    fast = fast.next;
  }

  if (fast === null) {
    head = head.next;
    return head;
  }

  while (fast.next !== null) {
    fast = fast.next;
    slow = slow.next;
  }

  slow.next = slow.next.next;

  return head;
};

const main = () => {
  let head = new ListNode(1);
  head.next = new ListNode(2);
  head.next.next = new ListNode(3);
  head.next.next.next = new ListNode(4);
  head.next.next.next.next = new ListNode(5);

  head = removeNthFromEnd(head, 2);

  curr = head

  while (curr !== null) {
    console.log(curr.val);
    curr = curr.next;
  }
}

if (require.main === module) {
  main();
}
